Make K8sPodOperator backwards compatible

dimberman opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

ashb commented on airflow/providers/cncf/kubernetes/backcompat/volume_mount.py:
```diff
@@ -0,0 +1,62 @@
+from kubernetes.client import models as k8s
+
+warnings.warn(
+    "This module is deprecated. Please use `kubernetes.client.models.V1Volume`.",
```
This needs to live as `from airflow.kubernetes.volume_mount import VolumeMount` to be a back-compat layer.

Same for all the others.

ashb replied:
Not the deprecation message, but this file. Calling it airflow.providers.cncf.kubernetes.backcompat.volume_mount won't help users.

dimberman replied:
@ashb  we can actually just change the import path using bowler for backport_providers. I was asked by @kaxil and @potiuk to set this up exclusively in the provider package. This will allow us to upgrade these as needed without releasing new airflow versions.

ashb replied:
See https://github.com/apache/airflow/pull/12384#discussion_r524534519

ashb commented on dev/provider_packages/refactor_provider_packages.py:
```diff
@@ -142,6 +142,7 @@ def rename_deprecated_modules(self) -> None:
             ("airflow.operators.bash", "airflow.operators.bash_operator"),
             ("airflow.operators.python", "airflow.operators.python_operator"),
             ("airflow.utils.session", "airflow.utils.db"),
+            ("airflow.providers.cncf.kubernetes.backcompat", "airflow.kubernetes"),
```
Oh - I just saw this -- that'll help us for backport, but won't help for anyone trying to use it on 2.0.0 (i.e. upgrade then fix errors approach)

dimberman committed:
Make the KubernetesPodOperator backwards compatible

This PR significantly reduces the pain of upgrading to Airflow 2.0
for users of the KubernetesPodOperator. Users will be allowed to
    continue using the airflow.kubernetes custom classes

potiuk reviewed (approved):
Great job @dimberman and @ashb . I think we are nicely handling all cases :). I think we might merge it just in time to release new backports :)

ashb commented on dev/provider_packages/refactor_provider_packages.py:
```diff
@@ -146,6 +146,13 @@ def rename_deprecated_modules(self) -> None:
+        def is_not_k8spodop(node: LN, capture: Capture, filename: Filename) -> bool:
+            return not filename.endswith("/kubernetes_pod.py")
+
+        self.qry.select_module("airflow.providers.cncf.kubernetes.backcompat").filter(
+            callback=is_not_k8spodop
+        ).rename("airflow.kubernetes")
```
Waaaaait -- just noticed that 1.10.x still has an `airflow/kubernetes/pod.py` etc. What is this going to do?

potiuk replied:
Yeah. They should be removed after rename. However Interesting thing is that the package installed correctly (prepare backport providers job in CI does it).

ashb replied:
Yeah, pip doesn't really care if two distributions install the same file -- it'll just happily overwrite the existing file.

Which is fine right until you come to uninstall things.

This brings up an interesting point. What is the behaviour when upgrading from 1.10.x with backport provider to 2.0.0 with normal provider. Yes pip will complain about version conflicts, but that still isn't an error (especially given many people will be on an older pip still)

potiuk replied:
From this respect good @ashb ! the backport provider package only contains airflow.providers files. We do not add airflow/kubernetes.

Just double-checked and installed both airflow 1.10.12 and backport cncf,kubernetes from pcakges and seem to be all ok.

This is what you get in 'backwards_compat_converters.py in the backport package:

```
from typing import List

from kubernetes.client import models as k8s

from airflow.exceptions import AirflowException
from airflow.kubernetes.pod import Port, Resources
from airflow.kubernetes.pod_runtime_info_env import PodRuntimeInfoEnv
from airflow.kubernetes.volume import Volume
from airflow.kubernetes.volume_mount import VolumeMount


def _convert_kube_model_object(obj, old_class, new_class):
    convert_op = getattr(obj, "to_k8s_client_obj", None)
    if callable(convert_op):
        return obj.to_k8s_client_obj()


```

ashb replied:
~What is this `rename()` achieving then? What dist does/should it appear in?~

potiuk replied:
One small thing that we could do - maybe also remove the backcompat classes. They seem to not be removed during the refactoring and they are not needed in the backport at all:

![image](https://user-images.githubusercontent.com/595491/99374141-91370280-28c2-11eb-8bc8-53b85bbeb601.png)

potiuk replied:
@ashb @dimberman Fixup here (I could not push it to astronomer's repo) : https://github.com/apache/airflow/pull/12392/commits/8d0be15ae4e4c1cd51d61ccd71fe0a8bef7f0ffe

potiuk replied:
> This brings up an interesting point. What is the behaviour when upgrading from 1.10.x with backport provider to 2.0.0 with normal provider. Yes pip will complain about version conflicts, but that still isn't an error (especially given many people will be on an older pip still)

This will work fine. Backport packages need airflow ~1.10. PIP will not allow to upgrade airflow without removing backports.

ashb replied:
(I think that's cos Dan didn't tick the "allow commits from maintainers?" Or maybe that only works for PRs from personal repos -- I had the same trying to push to a PolideaInternal PR you opened)

potiuk replied:
For example backport cncf.kubernetes:

```
       install_requires=[
            'apache-airflow>=1.10.12, <2.0.0',
            'cryptography>=2.0.0',
            'kubernetes>=3.0.0, <12.0.0',
        ],
```

potiuk replied:
We should simply add a step in UPDATING to 2.0 to make sure they remove backports simply (so that they are not tempted to force installation even with version conflict.

ashb replied:
> @ashb @dimberman Fixup here (I could not push it to astronomer's repo) : 8d0be15

Pushed this as https://github.com/apache/airflow/pull/12384/commits/d5d9e8b51b0ae534c8ac3f69a341c415ec8d3fb0

ashb replied:
@potiuk Turns out we need those files:

```
  File "/usr/local/lib/python3.6/site-packages/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py", line 39, in <module>
    from airflow.providers.cncf.kubernetes.backcompat.pod_runtime_info_env import PodRuntimeInfoEnv
ModuleNotFoundError: No module named 'airflow.providers.cncf.kubernetes.backcompat.pod_runtime_info_env'
```

ashb replied:
Force-pushed back to the previous commit.

potiuk replied:
But we should not I believe. This import should be changed in the backport provider as well.

potiuk replied:
So the proper fix will be to remove the classes from backcompat and add rename for that one as well.

potiuk replied:
the kubernetes_pod.py was excluded from the previous rename, but pod_runtime_info can be changed in all classes.

potiuk replied:
@dimberman ^^ have you seen this discussion ?

ashb commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -208,11 +221,15 @@ def __init__(  # pylint: disable=too-many-arguments,too-many-locals
         self.arguments = arguments or []
         self.labels = labels or {}
         self.startup_timeout_seconds = startup_timeout_seconds
-        self.env_vars = env_vars or []
+        self.env_vars = convert_env_vars(env_vars) if env_vars else []
+        if pod_runtime_info_envs:
+            self.env_vars.extend([convert_pod_runtime_info_env(p) for p in pod_runtime_info_envs])
         self.env_from = env_from or []
-        self.ports = ports or []
-        self.volume_mounts = volume_mounts or []
-        self.volumes = volumes or []
+        if configmaps:
+            self.env_from.extend([convert_configmap(c) for c in configmaps])
+        self.ports = [convert_port(p) for p in ports] if ports else []
+        self.volume_mounts = [convert_volume_mount(v) for v in volume_mounts] if volume_mounts else []
+        self.volumes = [convert_volume(volume) for volume in volumes] if volumes else []
```
These won't issue deprecation warnings, but that's okay because the imports will, right?

potiuk replied:
Right

potiuk commented on dev/provider_packages/refactor_provider_packages.py:
```diff
@@ -146,6 +146,13 @@ def rename_deprecated_modules(self) -> None:
+        self.qry.select_module("airflow.providers.cncf.kubernetes.backcompat").filter(
+            callback=is_not_k8spodop
+        ).rename("airflow.kubernetes")
+
```
```suggestion
        self.qry.select_module("airflow.providers.cncf.kubernetes.backcompat.pod_runtime_info")\
              .rename("airflow.kubernetes.pod_runtime_info")
```

That + restoring class removal should fix it.

ashb replied:
Does select_module just operate on that one module, not sub modules under it?

potiuk replied:
There are none under it :). It's the full package name not the module. But likely it should be select_package indeed :)

dimberman replied:
@potiuk @ashb wait can someone please break this down for me? I'm still confused as to what is broken.

potiuk replied:
@dimberman: Not broken, but can be improved :) 

This commit: https://github.com/apache/airflow/commit/8d0be15ae4e4c1cd51d61ccd71fe0a8bef7f0ffe removes the backcompat/pod.py and other files from backports as they are not needed there. Except that 

`from airflow.kubernetes.pod_runtime_info_env import PodRuntimeInfoEnv` in `kubernetes_pod.py` still points to backcompat because this file is excluded from the rename above.

The "perfect" fix is to:
1) Also replace the import in `./kubernetes_pod.py`. Can be likely done with this without affecting the other imports in ./kubernetes.pod:

```
        self.qry.select_package("airflow.providers.cncf.kubernetes.backcompat.pod_runtime_info")\
              .rename("airflow.kubernetes.pod_runtime_info")
```

2) remove the backcompat files as in https://github.com/apache/airflow/commit/8d0be15ae4e4c1cd51d61ccd71fe0a8bef7f0ffe

dimberman replied:
ohhh ok. I think I get it. 

So basically the idea here is that we want to delete the backcompat files in 1.10 since we're not using them, though we were running into a bowler issue because we weren't changing one import in `kubernetes_pod.py` (this was because we needed to keep the original import for the converters.  So adding that extra query fixes the `pod_runtime_info_env` import and allows us to delete those files.

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/backwards_compat_converters.py:
```diff
@@ -0,0 +1,123 @@
+    """
+    Converts an airflow Volume object into a k8s.V1Volume
+    @param volume:
+    @return: k8s.V1Volume
```
```suggestion
    :param volume:
    :return: k8s.V1Volume
```

Empty description for Volume

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/backwards_compat_converters.py:
```diff
@@ -0,0 +1,123 @@
+    """
+    Converts an airflow VolumeMount object into a k8s.V1VolumeMount
+    @param volume_mount:
+    @return: k8s.V1VolumeMount
```
same here

dimberman replied:
Is there a setting on black we can create for this or some form of a static check that automatically fixes this?

ashb replied:
No, black doesn't look at doc comments.

pydocstyle might.

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/backwards_compat_converters.py:
```diff
@@ -0,0 +1,123 @@
+def convert_resources(resources) -> k8s.V1ResourceRequirements:
+    """
+    Converts an airflow Resources object into a k8s.V1ResourceRequirements
+    @param resources:
```
same here

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/backwards_compat_converters.py:
```diff
@@ -0,0 +1,123 @@
+def convert_port(port) -> k8s.V1ContainerPort:
+    """
+    Converts an airflow Port object into a k8s.V1ContainerPort
+    @param port:
```
same here

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/backwards_compat_converters.py:
```diff
@@ -0,0 +1,123 @@
+def convert_env_vars(env_vars) -> List[k8s.V1EnvVar]:
+    """
+    Converts a dictionary into a list of env_vars
+    @param env_vars:
```
same here

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/backwards_compat_converters.py:
```diff
@@ -0,0 +1,123 @@
+def convert_pod_runtime_info_env(pod_runtime_info_envs) -> k8s.V1EnvVar:
+    """
+    Converts a PodRuntimeInfoEnv into an k8s.V1EnvVar
+    @param pod_runtime_info_envs:
```
same here

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/backwards_compat_converters.py:
```diff
@@ -0,0 +1,123 @@
+    """
+    Converts a PodRuntimeInfoEnv into an k8s.V1EnvVar
+    @param convert_image_pull_secrets:
+    @return:
```
same here

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/backwards_compat_converters.py:
```diff
@@ -0,0 +1,123 @@
+def convert_configmap(configmaps) -> k8s.V1EnvFromSource:
+    """
+    Converts a str into an k8s.V1EnvFromSource
+    @param configmaps:
```
same here

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/pod.py:
```diff
@@ -0,0 +1,110 @@
+        """
+        Converts to k8s object.
+
+        @rtype: object
```
```suggestion
        :rtype: object
```

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/pod_runtime_info_env.py:
```diff
@@ -0,0 +1,43 @@
+        self.field_path = field_path
+
+    def to_k8s_client_obj(self):
+        """:return: kubernetes.client.models.V1EnvVar"""
```
```suggestion
    def to_k8s_client_obj(self):
        """Converts to k8s object.
        
        :return: kubernetes.client.models.V1EnvVar
        """
```

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/pod_runtime_info_env.py:
```diff
@@ -0,0 +1,43 @@
+
+    def to_k8s_client_obj(self):
+        """:return: kubernetes.client.models.V1EnvVar"""
+        import kubernetes.client.models as k8s
```
Does this import need to be here anymore?

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/volume.py:
```diff
@@ -0,0 +1,64 @@
+    def __init__(self, name, configs):
+        """Adds Kubernetes Volume to pod. allows pod to access features like ConfigMaps
+        and Persistent Volumes
+        :param name: the name of the volume mount
```
```suggestion

        :param name: the name of the volume mount
```

kaxil commented on airflow/providers/cncf/kubernetes/backcompat/volume.py:
```diff
@@ -0,0 +1,64 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""This module is deprecated. Please use `airflow.providers.amazon.aws.hooks.datasync`."""
```
Copy/paste error: `datasync`

ashb commented on airflow/providers/cncf/kubernetes/backcompat/backwards_compat_converters.py:
```diff
@@ -0,0 +1,123 @@
+
+
+def convert_volume(volume) -> k8s.V1Volume:
+    """
+    Converts an airflow Volume object into a k8s.V1Volume
```
```suggestion
    """
    Converts an airflow Volume object into a k8s.V1Volume
    
```

Need a blank line between description and params

kaxil commented on kubernetes_tests/test_kubernetes_pod_operator_backcompat.py:
```diff
@@ -0,0 +1,948 @@
+    @mock.patch("airflow.kubernetes.pod_launcher.PodLauncher.monitor_pod")
+    @mock.patch("airflow.kubernetes.kube_client.get_kube_client")
+    def test_config_path(self, client_mock, monitor_mock, start_mock):  # pylint: disable=unused-argument
+        from airflow.utils.state import State
```
Can we move this import at top level -- since looks like it is used in more than 1 test

kaxil commented on kubernetes_tests/test_kubernetes_pod_operator_backcompat.py:
```diff
@@ -0,0 +1,948 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
```
How is this file testing anything differently than `kubernetes_tests/test_kubernetes_pod_operator.py` ??

kaxil replied:
Ok looks like you are importing VolumeMounts, Volume etc from the old paths to make sure it works

kaxil commented on kubernetes_tests/test_kubernetes_pod_operator_backcompat.py:
```diff
@@ -0,0 +1,948 @@
+    @mock.patch("airflow.kubernetes.pod_launcher.PodLauncher.start_pod")
+    @mock.patch("airflow.kubernetes.pod_launcher.PodLauncher.monitor_pod")
+    @mock.patch("airflow.kubernetes.kube_client.get_kube_client")
+    def test_config_path(self, client_mock, monitor_mock, start_mock):  # pylint: disable=unused-argument
```
This test does not test any old behaviour, does it?

i.e. it is just dupe of same test (`test_config_path`) in tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py

Can we remove all such tests that are not strictly checking the old behaviour too please?

dimberman committed:
spellcheck

dimberman committed:
spelling

dimberman committed:
clean up unecessary files in 1.10

dimberman committed:
clean up unecessary files in 1.10

dimberman committed:
clean up unecessary files in 1.10
